### ISSUE TITLE
Anyway to scrape by chunk?

### ISSUE THREAD

**Issue by junlin1998** [JustAnotherArchivist/snscrape#307]
Hi,

I am scraping a very large dataset using Snscrape （around 100 million). I used to use the get_items() function to iterate in python and load every tweet into the file, but now I think it would take a VERY long time if I am using iteration to get all the 100 million done.

Is there a way to scrape by chunk? I think I saw your answer in another closed thread that maybe we can scrape thousands at once. But I cannot figure it out by myself. Are there any functions or things?

Looking forward to your replies, and really appreciate your help!

Thank!

**Comment by JustAnotherArchivist** [comment]
snscrape already does that, namely in chunks of 100 tweets (which, I believe, was the limit at the time I wrote the code).

But you're right, such large scrapes take a very long time. One thing you can do is split it up into multiple independent scrapes and run those in parallel. For example, you might run one scrape per calendar month. Instead of `snscrape --jsonl twitter-search foo`, you'd run `snscrape --jsonl twitter-search 'foo since:2021-01-01 until:2021-02-01'` etc. for every month of interest. (Note that `since:` is inclusive and `until:` is exclusive, i.e. that command fetches all matching tweets posted in January 2021.) Depending on how many results you expect, you might want to run even smaller time windows, down to individual days.

Parallelism inside snscrape isn't something I intend to add at this time as it adds significant complexity, so I suggest doing this externally using a little shell script and `xargs` or `parallel` instead.

For example, this script takes a search query and a month as arguments (i.e. `./twitter-search-month foo 2021-01` is equivalent to the example above) and produces an output file and a log file for that month in the current directory:

```bash
#!/bin/bash
query="$1"
month="$2"
snscrape --jsonl --verbose twitter-search "${query} since:${month}-01 until:$(date --date="${month}-01 + 1 month" +%Y-%m-%d)" >"${month}.jsonl" 2>"${month}.log"
```

You can then generate a list of months and execute that; e.g. to scrape the whole of 2020 for 'foo' using `parallel`: `echo 2020-{01..12} | tr ' ' '\n' | parallel -P 12 ./twitter-search-month foo`

(I'm not sure how portable the `date` command is. I'm using the GNU coreutils version. But this is mostly intended to illustrate one possible way to approach this problem, not a direct solution that'll work for everyone without changes.)

**Comment by junlin1998** [comment]
Hi, thank you so much for your reply. It's good to know that there is no built-in way for snscrape to do the scraping. Honestly, I am doing that by separating the date range into individual days. BTW, I am thinking of using multi-processing/treading but I am really new to these notions. One concern is, would running multiple scrapers at one time on one machine (i.e., one local network) cause a higher likelihood to get my IP banned?

Again, thank you so much for your great work and your replies. Much appreciated!

**Comment by JustAnotherArchivist** [comment]
In general, it certainly would, but I'm not sure how strict their limits are. As an anecdotal data point, I've run 8‒10 scrapes in parallel without issues in the past. That may or may not still be true.

**Comment by JustAnotherArchivist** [comment]
A note on multithreading: snscrape is currently not verified to be thread-safe as that wasn't a design goal. I can't think of any reason right now why it wouldn't be (logging is, and I don't think there are any other non-constant objects that would be shared between threads), but it's not guaranteed. I'd recommend avoiding it.

**Comment by junlin1998** [comment]
THANK YOU! I'm done w questions, let me close it for you
